### PR TITLE
Add Group support

### LINKED ISSUE #57: Add groups support
We got the annotation value, but we're not using it. I'm thinking we can extract the value in the adapters itself and skip validation if the validation request doesn't contain the correct groups.

### PR DESCRIPTION
**Pull request by SentryMan**
- adds group varargs to validate methods 
- adds groups to ValidationRequest
- Adds new default method to validation adapter to check if groups match
- Adapters modified to check groups before validating
- adds a failfast mode

fixes #57

### CODE REVIEW

File: validator/src/main/java/io/avaje/validation/core/DValidator.java [hunk 1]
```diff
@@ -53,22 +56,23 @@ final class DValidator implements Validator, ValidationContext {
...
-  public void validate(Object any) {
-    validate(any, null);
+  public void validate(Object any, @Nullable Class<?>... groups) {
+    validate(any, null, groups);
   }
 
   @Override
   @SuppressWarnings("unchecked")
-  public void validate(Object any, @Nullable Locale locale) {
+  public void validate(Object any, @Nullable Locale locale, @Nullable Class<?>... groups) {
```
Comment (rbygrave) [inline validator/src/main/java/io/avaje/validation/core/DValidator.java:73]: Is groups Nullable?  or just an empty array

File: validator/src/main/java/io/avaje/validation/Validator.java [hunk 2]
```diff
@@ -12,12 +13,13 @@
...
 import java.util.ResourceBundle;
 import java.util.ServiceLoader;
+import java.util.Set;
 import java.util.function.Supplier;
 
 public interface Validator {
 
   /** Validate the object using the default locale. */
-  void validate(Object any) throws ConstraintViolationException;
+  void validate(Object value, @Nullable Class<?>... groups) throws ConstraintViolationException;
```
Comment (rbygrave) [inline validator/src/main/java/io/avaje/validation/Validator.java:22]: With varargs groups I don't think it's nullable ?

File: validator/src/main/java/io/avaje/validation/core/CoreAdapterBuilder.java [hunk 3]
```diff
@@ -73,12 +76,18 @@ <T> ValidationAdapter<T> build(Type type, Object cacheKey) {
...
         return result;
       }
     }
     // unknown annotations have noop
     return NOOP;
   }
+
+  static Set<Class<?>> extractGroups(Map<String, Object> attributes) {
+    @SuppressWarnings("unchecked")
+    final var annotationsGroups = (List<Class<?>>) attributes.getOrDefault("groups", List.of());
```
Comment (rbygrave) [inline validator/src/main/java/io/avaje/validation/core/CoreAdapterBuilder.java:90]: At generation time we could put immutable set of class into the attributes. Then we could skip the list -> set conversion here ?

**Review by rbygrave — APPROVED** [review]

File: validator/src/main/java/io/avaje/validation/Validator.java (hunk 2 above)
Comment (SentryMan) [reply validator/src/main/java/io/avaje/validation/Validator.java:22]: If you don't pass anything for varargs it's null

File: validator/src/main/java/io/avaje/validation/core/CoreAdapterBuilder.java (hunk 3 above)
Comment (SentryMan) [reply validator/src/main/java/io/avaje/validation/core/CoreAdapterBuilder.java:90]: I could have done that, but I forgot

**Pull request merged**